client/db/bolt/upgrades.go

```diff
@@ -130,110 +132,207 @@ func getVersionTx(tx *bbolt.Tx) (uint32, error) {
 	return intCoder.Uint32(versionB), nil
 }
 
-func v1Upgrade(dbtx *bbolt.Tx) error {
+func v1Upgrade(db *bbolt.DB) error {
 	const oldVersion = 0
 
-	if err := ensureVersion(dbtx, oldVersion); err != nil {
+	if err := ensureVersion(db, oldVersion); err != nil {
 		return err
 	}
 
-	bkt := dbtx.Bucket(appBucket)
-	if bkt == nil {
-		return fmt.Errorf("appBucket not found")
-	}
-
-	return reloadMatchProofs(dbtx)
+	return reloadMatchProofs(db)
 }
 
 // v2Upgrade adds a MaxFeeRate field to the OrderMetaData. The upgrade sets the
 // MaxFeeRate field for all historical orders to the max uint64. This avoids any
 // chance of rejecting a pre-existing active match.
-func v2Upgrade(dbtx *bbolt.Tx) error {
+func v2Upgrade(db *bbolt.DB) error {
 	const oldVersion = 1
 
-	dbVersion, err := fetchDBVersion(dbtx)
-	if err != nil {
-		return fmt.Errorf("error fetching database version: %w", err)
+	if err := ensureVersion(db, oldVersion); err != nil {
+		return err
 	}
 
-	if dbVersion != oldVersion {
-		return fmt.Errorf("v2Upgrade inappropriately called")
-	}
+	fmt.Println("Adding max fee rates to orders in the database.  This may take a while...")
+	start := time.Now()
 
 	// For each order, set a maxfeerate of max uint64.
 	maxFeeB := uint64Bytes(^uint64(0))
 
-	master := dbtx.Bucket(ordersBucket)
-	if master == nil {
-		return fmt.Errorf("failed to open orders bucket")
+	// doBatch contains the primary logic for updating max fee in batches.
+	// This is done because attempting to migrate in a single database
+	// transaction could result in massive memory usage and could potentially
+	// crash on many systems due to ulimits.
+	const maxEntries = 20000
+	var (
+		resumeOffset uint32
+		numUpdated   uint32
+		totalUpdated uint64
+	)
+	doBatch := func() error {
+		return db.Update(func(tx *bbolt.Tx) error {
+			master := tx.Bucket(ordersBucket)
+			if master == nil {
+				return fmt.Errorf("failed to open orders bucket")
+			}
+			numUpdated = 0
+			var numIterated uint32
+			return master.ForEach(func(oid, _ []byte) error {
+
+				if numUpdated >= maxEntries {
+					return batchDoneErr
+				}
+
+				// Skip entries that have already been migrated in previous batches.
+				numIterated++
+				if numIterated-1 < resumeOffset {
+					return nil
+				}
+				resumeOffset++
+
+				oBkt := master.Bucket(oid)
+				if oBkt == nil {
+					return fmt.Errorf("order %x bucket is not a bucket", oid)
+				}
+				numUpdated++
+				return oBkt.Put(maxFeeRateKey, maxFeeB)
+			})
+		})
 	}
 
-	return master.ForEach(func(oid, _ []byte) error {
-		oBkt := master.Bucket(oid)
-		if oBkt == nil {
-			return fmt.Errorf("order %x bucket is not a bucket", oid)
+	// Update all entries in batches.
+	for {
+		err := doBatch()
+		totalUpdated += uint64(numUpdated)
+		if err != nil {
+			if errors.Is(err, batchDoneErr) {
+				fmt.Printf("Updated %d entries (%d total)\n", numUpdated, totalUpdated)
+				continue
+			}
+			return err
```

Comment:
If we hit here, that means the user has to restore their db backup right?  Perhaps a message.